### PR TITLE
Support react16 new render return types

### PR DESCRIPTION
**Pull request by gwuhaolin**
See react16 doc:
https://facebook.github.io/react/docs/react-component.html#render

Please fill in this template.

- [ok] Use a meaningful title for the pull request. Include the name of the package modified.
- [ok] Test the change in your own code. (Compile and run.)
- [ok] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ok] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ok] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [ok] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.

### CODE REVIEW

File: types/react/index.d.ts [hunk 1]
```diff
@@ -282,7 +282,7 @@ declare namespace React {
         // tslint:enable:unified-signatures
 
         forceUpdate(callBack?: () => any): void;
-        render(): JSX.Element | null | false;
+        render(): JSX.Element | JSX.Element[] | string | number | null | false;
```
Comment (vsiao) [inline types/react/index.d.ts]: Maybe `React.ReactNode` is more accurate (though I'm not sure how that will interact with JSX). `JSX.Element[]` doesn't capture the fact that the array is allowed to contain non-element values.

**Review by vsiao — CHANGES_REQUESTED** [review]

**Comment by Kovensky** [comment]
This seems related to https://github.com/DefinitelyTyped/DefinitelyTyped/pull/19363

**Comment by apexskier** [comment]
I agree with @vsiao about `ReactNode` being more appropriate (though I haven't tested it with react 16. I assume `boolean` returns are okay). [Stateless components also need an updated return type](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/424349146aee831c31ee1823e968c40a0a65aaba/types/react/index.d.ts#L320)

**Comment by RyanCavanaugh** [comment]
@gwuhaolin Please fix the failures indicated in the Travis CI log.

**Comment by tkrotoff** [comment]
Would also be nice to add a test

**Comment by Pajn** [comment]
As previously said, it should be `ReactNode` instead. I can confirm that `boolean` returns are okay.

File: types/react/index.d.ts (hunk 1 above)
Comment (Magellol) [reply types/react/index.d.ts]: Maybe adding `React.ReactNode` would fix the following pattern. I'm currently getting an error in TS right now but I believe React allows us to have one item that is `null` to remove the node.

```jsx
render() {
  return [
    <button key="modal-button" onClick={this.openModal}>
        Open modal
    </button>,
    this.state.shouldRenderModal ? (
      <Modal />
    ) : null,
  ];
}
```

As a temporary fix I can do

```diff
render() {
  return [
    <button key="modal-button" onClick={this.openModal}>
        Open modal
      </button>,
-      this.state.shouldRenderModal ? (
+      (this.state.shouldRenderModal ? (
       <Modal />
-      ) : null,
+      ) : null) as JSX.Element,
  ];
}
```

**Comment by minestarks** [comment]
I'm closing this PR since it has test failures that haven't been addressed in over a month. If you'd still like to get these changes in, please open a new PR. Thank you!

**Pull request closed without merging**